File: restful-api/src/main/java/kr/co/restfulapi/rest/ProductController.java
package kr.co.restfulapi.rest;

import kr.co.restfulapi.dto.MemberDto;
import kr.co.restfulapi.dto.ProductCreateDto;
import kr.co.restfulapi.dto.ProductDto;
import kr.co.restfulapi.dto.SignUpRequest;
import kr.co.restfulapi.service.MemberService;
import kr.co.restfulapi.service.ProductService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;

@Slf4j
@RestController
@RequestMapping("/product")
@RequiredArgsConstructor
public class ProductController {

	private final ProductService productService;

	@GetMapping("/{productId}")
	public ResponseEntity<ProductDto> findProduct(@PathVariable Long productId) {
		ProductDto productDto = productService.findById(productId);
		return ResponseEntity.ok(productDto);
	}

    @PostMapping
	public ResponseEntity<ProductDto> saveProduct(@Valid ProductCreateDto request) {
		ProductDto productDto = productService.createProduct(request);
		return ResponseEntity.ok(productDto);
	}

}

File: store-min/product-api/src/main/kotlin/com/store/productapi/controller/ProductController.kt
package com.store.productapi.controller

import com.store.core.domain.product.ProductCreateDomain
import com.store.core.domain.product.ProductDomain
import com.store.core.http.ApiResponse
import com.store.productapi.dto.ProductResponseDto
import com.store.productapi.service.ProductService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/product-kotlin")
class ProductController(private val productService: ProductService) {

    @GetMapping
    fun getProducts(): ResponseEntity<ApiResponse> {
        val productDomains: List<ProductDomain> = productService.getProducts()
        val productResponseDtos: List<ProductResponseDto> = productDomains.map { ProductResponseDto.from(it) }

        return ResponseEntity.ok()
            .body(ApiResponse.success(productResponseDtos))
    }

    @GetMapping("/{id}")
    fun getProduct(@PathVariable id: Long): ResponseEntity<ApiResponse> {
        val productDomain: ProductDomain = productService.getById(id)
        return ResponseEntity.ok()
            .body(
                ApiResponse
                    .success(ProductResponseDto.from(productDomain))
            )
    }

    @PostMapping
    fun register(@RequestBody productCreateDomain: ProductCreateDomain): ResponseEntity<ApiResponse> {
        val productDomain: ProductDomain = productService.register(productCreateDomain)
        return ResponseEntity.ok()
            .body(
                ApiResponse
                    .success(ProductResponseDto.from(productDomain))
            )
    }
}
